internal/controller/runner.go

```diff
@@ -0,0 +1,226 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	mcadv1beta1 "github.com/project-codeflare/mcad/api/v1beta1"
+)
+
+// AppWrapperReconciler responsible for the portion of the lifecycle that happens on the execution cluster
+type Runner struct {
+	AppWrapperReconciler
+}
+
+// permission to edit appwrappers
+
+//+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers/finalizers,verbs=update
+
+// permission to edit wrapped resources: pods, services, jobs, podgroups
+
+//+kubebuilder:rbac:groups="",resources=pods;services,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apps,resources=deployments;statefulsets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=scheduling.sigs.k8s.io,resources=podgroups,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=scheduling.x-k8s.io,resources=podgroups,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kubeflow.org,resources=pytorchjobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.ray.io,resources=rayclusters,verbs=get;list;watch;create;update;patch;delete
+
+// permission to view nodes
+
+//+kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch
+
+// Reconcile one Running AppWrapper on an execution cluster and monitor health of its execution resources.
+//
+//gocyclo:ignore
+func (r *Runner) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// get deep copy of AppWrapper object in reconciler cache
+	appWrapper := &mcadv1beta1.AppWrapper{}
+	if err := r.Get(ctx, req.NamespacedName, appWrapper); err != nil {
+		// no such AppWrapper, nothing to reconcile, not an error
+		return ctrl.Result{}, nil
+	}
+
+	// append appWrapper ID to logger
+	ctx = withAppWrapper(ctx, appWrapper)
+
+	// abort and requeue reconciliation if reconciler cache is stale
+	if r.isStale(ctx, appWrapper) {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// handle deletion
+	if !appWrapper.DeletionTimestamp.IsZero() {
```

Comment:
Maybe also check that the runner finalizer is still there?

Reply:
done